File: android/DFM/src/main/java/com/dontflymoney/language/Language.java
package com.dontflymoney.language;

import java.util.Locale;

import android.app.Activity;
import android.content.res.Configuration;
import android.content.res.Resources;

import com.dontflymoney.baseactivity.SmartActivity;
import com.dontflymoney.io.File;
import com.dontflymoney.io.FileNames;


public class Language
{
	public static void ChangeAndSave(SmartActivity activity, String language)
	{
		String current = Locale.getDefault().toString();
		
		language = language.replace("-", "_");
		
		if (language.equalsIgnoreCase(current))
			return;

		change(activity, language);
		
		File file = new File(activity, FileNames.Language);
		file.WriteToFile(language);
		
		activity.refresh();
	}
	
	public static void ChangeFromSaved(Activity activity)
	{
		File file = new File(activity, FileNames.Language);
		
		String language = file.ReadFromFile();
		
		change(activity, language);
	}

	private static void change(Activity activity, String language)
	{
		Resources resources = activity.getResources();

		Locale[] locales = Locale.getAvailableLocales();
		Locale locale = null;
		
		for(int l = 0; l < locales.length; l++)
		{
			if (locales[l].toString().equalsIgnoreCase(language))
			{
				locale = locales[l]; 
			}
		}
		
		if (locale == null)
			return;
		
		Locale.setDefault(locale);

		Configuration config = resources.getConfiguration();
		config.locale = locale;
		
		resources.updateConfiguration(config, null);
		resources.flushLayoutCache();
	}
	
}

File: android/DFM/src/main/kotlin/com/darakeon/user/Language.kt
package com.dontflymoney.userdata

import android.app.Activity
import com.dontflymoney.activityObjects.SmartStatic
import com.dontflymoney.baseactivity.SmartActivity
import java.util.*


object Language {
    private val spKey = "Language"

    fun <T : SmartStatic> ChangeAndSave(activity: SmartActivity<T>, language: String) {
        var language = language
        val current = Locale.getDefault().toString()

        language = language.replace("-", "_")

        if (language.equals(current, ignoreCase = true))
            return

        change(activity, language)

        SP.setValue(activity, spKey, language)

        activity.refresh()
    }

    fun ChangeFromSaved(activity: Activity) {
        val language = SP.getValue(activity, spKey)

        if (language != null)
            change(activity, language)
    }

    private fun change(activity: Activity, language: String) {
        val resources = activity.resources

        val availableLocales = Locale.getAvailableLocales()
        var locale: Locale? = null

        for (availableLocale in availableLocales) {
            if (availableLocale.toString().equals(language, ignoreCase = true)) {
                locale = availableLocale
            }
        }

        if (locale == null)
            return

        Locale.setDefault(locale)

        val config = resources.configuration
        config.setLocale(locale)

        //NEW
        activity.createConfigurationContext(config)

        //OLD
        resources.updateConfiguration(config, null)
        resources.flushLayoutCache()
    }

}